### PR TITLE
API Enable chained injector dependencies

### PR DESCRIPTION
**Pull request by tractorcow**
This feature will allow more advance configuration of changed dependencies (especially if chained across modules).

### CODE REVIEW

File: control/injector/Injector.php [hunk 1]
```diff
@@ -543,14 +543,25 @@ protected function instantiate($spec, $id=null, $type = null) {
...
 		$constructorParams = array();
 		if (isset($spec['constructor']) && is_array($spec['constructor'])) {
 			$constructorParams = $spec['constructor'];
 		}
 
-		$factory = isset($spec['factory']) ? $this->get($spec['factory']) : $this->getObjectCreator();
-		$object = $factory->create($class, $constructorParams);
+		// Catch service references
+		if (strpos($class, '%$') === 0) {
+			$otherID = substr($class, 2);
```
Comment (tractorcow) [inline control/injector/Injector.php]: At a higher level the "raw" dependency config is stuck in an `array('class' => $spec)` so I've opted to keep it simple here and just check the classname.

**Comment by sminnee** [comment]
Can we have some docs please?

**Comment by sminnee** [comment]
I don't think that random tweaks to Injector are a good idea to merge. Can we get a bit more detail about why this is needed, an amendment to the docs that describes that use-case, and a test that covers that use-case?

**Comment by tractorcow** [comment]
This is the typical use case where this is useful:

```yaml
---
Name: assetstore
---
Injector:
  FlysystemAssetStore:
    class: 'SilverStripe\Filesystem\Flysystem\FlysystemAssetStore'
    properties:
      Filesystem: '%$FlysystemBackend'
  # Assign as default story
  AssetStore: '%$FlysystemAssetStore'
```

Given this config, without this code trying to create an instance of `AssetStore` via injector gives the following error.

```
Class %$FlysystemAssetStore does not exist
ReflectionClass->__construct(%$FlysystemAssetStore)
InjectionCreator.php:14

InjectionCreator->create(%$FlysystemAssetStore,Array)
Injector.php:563

Injector->instantiate(Array,AssetStore)
Injector.php:892

Injector->get(AssetStore)
DBFileTest.php:5

--------------------
```

I could construct my yml this way


```yaml
Injector:
  AssetStore:
    class: 'SilverStripe\Filesystem\Flysystem\FlysystemAssetStore'
    properties:
      Filesystem: '%$FlysystemBackend'
  ```
  
  But now to override it with a store that doesn't use flysystem, I'd need to do something like the below:
  
  
```yaml
Injector:
  AssetStore:
    class: 'SomeOtherClass'
    properties:
      Filesystem: null
```

Whereas with this code, I could simply do

```yaml
Injector:
  AssetStore: '%$SomeOtherService'
```

**Comment by tractorcow** [comment]
Documentation with another example added, and I've updated and renamed the test.

**Comment by sminnee** [comment]
Hrm, I think that the underlying issue here is probably this part:

> But now to override it with a store that doesn't use flysystem, I'd need to do something like the below

It's too hard to override a module-default configuration. I'm a bit worried if we're expecting that all module-default injector configurations need to use the chaining syntax to avoid this issue.

Injector probably needs to provide an easier way of replacing an entire service definition, something like this:

    Injector:
      AssetStore[replace]:
        class: 'SomeOtherClass'

I'm not convinced that it's a good idea to merge this change. I'm not expecting that its absence should block the filesystem work, should it, @tractorcow?

I think it would be better to think through a few limitations of Injector and propose a more broadly set of changes to it for SS4. Otherwise we risk chasing our tail.

**Comment by tractorcow** [comment]
We already discussed merge strategies for config a while back; I think our conclusion is that it was possible but it would have more performance impact.

I don't think chaining is necessarily a poor pattern, since it only adds flexibility, and doesn't actually oblige users to utilise it. It solves the same problem that php variables do; An assigned value can be delegated to another variable (or passed to a method). This feature just makes the configuration we define a lot more transportable. Is that the right word?

Without it, users need to be careful about how they assign configs for services, since it could easily be done in such a way as to make it difficult to override. Up until now we haven't had this problem, but now that we are defining complex configuration for things such as logging and asset abstraction, it's going to be necessary to change something.

**Comment by sminnee** [comment]
It would be good to get some other views here—I'm willing to be swayed on this one. ;-)

**Comment by stevie-mayhew** [comment]
I would like to see a way to over ride configuration values rather than merge - we run into the problem quite a lot when working with other developers modules (or our own :smile:)

A common use case is where we have to use the PHP configuration layer to do something like the following to remove a tab from a model admin.

```php
Config::inst()->remove('ShopAdmin', 'managed_models');
Config::inst()->update('ShopAdmin', 'managed_models', array(
    'Order',
    'ShopConfig'
));
```

This would be better as something like 
```
Config::inst()->replace('ShopAdmin', 'managed_models', array(
    'Order',
    'ShopConfig'
));
```

And a YML layer configuration such as the one described by @sminnee would be a good move as well, IMHO. What happens if you are using a module which defines a default asset store, but you want to have another one running in parallel? Or replace the default store with another, then add a third?

**Comment by tractorcow** [comment]
I want both those features AND this one. :)

**Comment by tractorcow** [comment]
Closing this PR since it doesn't seem to be going anywhere.

**Pull request closed without merging**